File: sse.go
package apptheory

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"strings"
)

// SSEEvent is a Server-Sent Events (SSE) message.
//
// Framing rules:
// - Optional: "id: <id>\n"
// - Optional: "event: <event>\n"
// - Data: one "data: <line>\n" per line (at least one line)
// - Terminator: "\n"
type SSEEvent struct {
	ID    string
	Event string
	Data  any
}

func sseDataString(value any) (string, error) {
	if value == nil {
		return "", nil
	}
	switch v := value.(type) {
	case string:
		return v, nil
	case []byte:
		return string(v), nil
	default:
		b, err := json.Marshal(v)
		if err != nil {
			return "", fmt.Errorf("apptheory: sse json marshal: %w", err)
		}
		return string(b), nil
	}
}

func formatSSEEvent(event SSEEvent) ([]byte, error) {
	var buf bytes.Buffer

	id := strings.TrimSpace(event.ID)
	if id != "" {
		buf.WriteString("id: ")
		buf.WriteString(id)
		buf.WriteByte('\n')
	}

	name := strings.TrimSpace(event.Event)
	if name != "" {
		buf.WriteString("event: ")
		buf.WriteString(name)
		buf.WriteByte('\n')
	}

	data, err := sseDataString(event.Data)
	if err != nil {
		return nil, err
	}

	data = strings.ReplaceAll(data, "\r\n", "\n")
	data = strings.ReplaceAll(data, "\r", "\n")
	lines := strings.Split(data, "\n")
	if len(lines) == 0 {
		lines = []string{""}
	}
	for _, line := range lines {
		buf.WriteString("data: ")
		buf.WriteString(line)
		buf.WriteByte('\n')
	}

	buf.WriteByte('\n')
	return buf.Bytes(), nil
}

// SSEResponse builds a canonical AppTheory Response with properly framed SSE output.
func SSEResponse(status int, events ...SSEEvent) (*Response, error) {
	var buf bytes.Buffer
	for _, ev := range events {
		b, err := formatSSEEvent(ev)
		if err != nil {
			return nil, err
		}
		buf.Write(b)
	}
	return &Response{
		Status: status,
		Headers: map[string][]string{
			"content-type":  {"text/event-stream"},
			"cache-control": {"no-cache"},
			"connection":    {"keep-alive"},
		},
		Cookies:  nil,
		Body:     buf.Bytes(),
		IsBase64: false,
	}, nil
}

// SSEStreamResponse builds a canonical AppTheory Response with event-by-event SSE output.
//
// The returned response uses response streaming when invoked through the API Gateway REST API v1 adapter
// (`ServeAPIGatewayProxy` via `HandleLambda`).
func SSEStreamResponse(ctx context.Context, status int, events <-chan SSEEvent) (*Response, error) {
	if ctx == nil {
		ctx = context.Background()
	}

	pr, pw := io.Pipe()
	go func() {
		defer pw.Close()

		if events == nil {
			return
		}

		for {
			select {
			case <-ctx.Done():
				return
			case ev, ok := <-events:
				if !ok {
					return
				}

				b, err := formatSSEEvent(ev)
				if err != nil {
					_ = pw.CloseWithError(err)
					return
				}

				if _, err := pw.Write(b); err != nil {
					_ = pw.CloseWithError(err)
					return
				}
			}
		}
	}()

	return &Response{
		Status: status,
		Headers: map[string][]string{
			"content-type":  {"text/event-stream"},
			"cache-control": {"no-cache"},
			"connection":    {"keep-alive"},
		},
		Cookies:    nil,
		Body:       nil,
		BodyReader: pr,
		IsBase64:   false,
	}, nil
}

// MustSSEResponse builds an SSE response and panics on framing/serialization errors.
func MustSSEResponse(status int, events ...SSEEvent) *Response {
	resp, err := SSEResponse(status, events...)
	if err != nil {
		panic(err)
	}
	return resp
}

File: py/src/apptheory/sse.py
from __future__ import annotations

import json as jsonlib
from dataclasses import dataclass
from typing import Any, Iterable, Iterator

from apptheory.response import Response, normalize_response


@dataclass(slots=True)
class SSEEvent:
    id: str = ""
    event: str = ""
    data: Any = ""

    def __init__(self, *, id: str = "", event: str = "", data: Any = "") -> None:
        self.id = str(id or "").strip()
        self.event = str(event or "").strip()
        self.data = data


def _sse_data_string(value: Any) -> str:
    if value is None:
        return ""
    if isinstance(value, bytes):
        return value.decode("utf-8", errors="replace")
    if isinstance(value, str):
        return value
    return jsonlib.dumps(value, separators=(",", ":"), ensure_ascii=False, sort_keys=True)


def format_sse_event(event: SSEEvent) -> bytes:
    parts: list[str] = []

    if event.id:
        parts.append(f"id: {event.id}\n")
    if event.event:
        parts.append(f"event: {event.event}\n")

    data = _sse_data_string(event.data).replace("\r\n", "\n").replace("\r", "\n")
    lines = data.split("\n") if data is not None else [""]
    if not lines:
        lines = [""]
    for line in lines:
        parts.append(f"data: {line}\n")

    parts.append("\n")
    return "".join(parts).encode("utf-8")


def sse(status: int, events: list[SSEEvent]) -> Response:
    framed = b"".join([format_sse_event(e) for e in (events or [])])
    return normalize_response(
        Response(
            status=int(status or 200),
            headers={
                "content-type": ["text/event-stream"],
                "cache-control": ["no-cache"],
                "connection": ["keep-alive"],
            },
            cookies=[],
            body=framed,
            is_base64=False,
        )
    )


def sse_event_stream(events: Iterable[SSEEvent] | None) -> Iterator[bytes]:
    for event in events or []:
        yield format_sse_event(event)
